modules/README.md

```diff
@@ -115,35 +119,235 @@ Kudos to @mfw78
+
+This allows us to add additional logic when validating signatures, such as differentiating the validation logic based on the message. The challenge is that it is impossible to retrieve the message that was signed from the hash. To work around it we can encode the message into the signature and then when the smart contract is called this message is extracted, hashed and validated against the hash. This is outlined in the following example code:
+
+```solidity
+function isValidSignature(bytes32 hash, bytes memory payload) external view returns (bytes4) {
+  (bytes memory signature, bytes memory message) = abi.decode(payload, (bytes,bytes));
+  require(hash == keccak256(message), "Unexpected message/hash combination");
+  // more validation logic
+}
+```
+
+Using this approach it is also possible to encode ERC-712 based messages in the signature (for signatures requested via `eth_signTypedData`). With this we can route the signature validation based on the domain of the ERC-712 message. The payload contain the signature, the domainHash and the hashStruct. Only minor adjustments are required to the previous code:
+
+```solidity
+mapping(bytes32 => ISignatureValidator) validatorForDomain;
+function isValidSignature(bytes32 hash, bytes memory payload) external view returns (bytes4) {
+  (bytes32 domainHash, bytes32 hashStruct, bytes memory signature) = abi.decode(payload, (bytes32,bytes32,bytes));
+  require(hash, keccak256(0x19, 0x01, domainHash, hashStruct), "Could not validate data");
+  ISignatureValidator validator = validatorForDomain[domainHash];
+  return validator.isValidSignature(msg.sender /* account */, _msgSender(), domainHash, hashStruct, signature);
+}
+```
+
+To distinguish signatures that follow this format from "default" signatures the signature is prepended with a special bytes4 identifier. When the dapp requests the signature the wallet can then create the signature in the required format for the correct signature routing.
+
+The encoded data to be received by the signature validator manager as `bytes` parameter of `isValidSignature(bytes32,bytes)` is expected to be encoded as follows:
+
+```solidity
+function encodeData(bytes4 selector, bytes32 domain, bytes32 hashStruct, bytes calldata signatures) public pure returns (bytes memory) {
+        bytes memory data = abi.encode(domain, hashStruct, signatures);
+        return abi.encodePacked(selector, data);
+}
+```
+
+### SignatureValidatorManager with SafeProtocolManager
+
+An account must set `SafeProtocolManager` as a fallback handler and set `SignatureValidatorManager` as function handler for [isValidSignature(bytes32,bytes)](https://eips.ethereum.org/EIPS/eip-1271) function in the manager. So, when a contract wants to verify account signature, it should call `isValidSignature(bytes32,bytes)` on the account and the `SafeProtocolManager` will forward the call to the `SignatureValidatorManager` which will call the `SignatureValidatorContract` to check if the signature is valid.
```

Comment:
This adds quite some gas overhead, is there a way to optimize this?

Reply:
Potentially it makes sense to not mention anything about `ISafeProtocolSignatureValidatorManager.handle` and how this is set. For the protocol in itself this is not really important. How the validation  is routed and the interface of the validator or the only necessary facts for the spec, the other thing is a implementation detail.

That being said, I think we need to iterate on what is part of the Manager and what is delegated to other contracts.

Reply:
If we remove these details, the whole section of `SignatureValidatorManager with SafeProtocolManager` will be removed and it is not clear how `isValidSignature(bytes32,bytes)` is connected to validation flow